src/quorumnet/bt_serialize.h

```diff
@@ -0,0 +1,603 @@
+// Copyright (c)      2019, The Loki Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+#include <iostream> // DEBUG
+
+// Please no epee.  *Especially* for serialization!
+#include <string>
+#ifdef __cpp_lib_string_view
+#include <string_view>
+#endif
+#include <vector>
+#include <list>
+#include <unordered_map>
+#include <algorithm>
+#include <functional>
+#include <ostream>
+#include <sstream>
+#include <streambuf>
+#include <boost/variant.hpp>
+#include "common.h"
+
+#include <iostream> // DEBUG
+
+namespace quorumnet {
+
+/** \file
+ * Quorumnet serialization is very simple: we support two primitive types, strings and integers,
+ * and two container types, lists and dicts with string keys.  On the wire these go in BitTorrent
+ * byte encoding as described in BEP-0003 (https://www.bittorrent.org/beps/bep_0003.html#bencoding).
+ *
+ * On the C++ side, on input we allow strings, integral types, STL-like containers of these types,
+ * and STL-like containers of pairs with a string first value and any of these types as second
+ * value.  We also accept boost::variants and std::variants (if compiled with std::variant support,
+ * i.e. in C++17 mode) that contain any of these.
+ *
+ * One minor deviation from BEP-0003 is that we don't support serializing values that don't fit in a
+ * 64-bit integer.
```

Comment:
It can serialize a u64, but no larger.  (Though that's just an implementation detail; in theory we could support larger without too much trouble but I don't see a point in implementing it at the moment).

Reply:
Does this mean bt_value needs to include uint64_t in the variant declaration? But I also think that if we intentionally restrict it to i64 it covers all our use cases and simplifies implementation.

Reply:
No, the `int64_t` holds anything you might want to put into it or whatever size, including high-bit `uint64_t` values; they need to be `static_cast` back to a `uint64_t` to get them out, but calling `get_int<T>(val)` does that (along with range checking on T for any smaller types).